replay2/memory_remapper/memory_remapper.cc

```diff
@@ -14,4 +14,98 @@
 
 #include "memory_remapper.h"
 
-void foo(int var) { return; }
+namespace agi {
+namespace replay2 {
+
+	ReplayAddress MemoryRemapper::AddMapping(const MemoryObservation& observation)
+	{
+		const CaptureAddress captureAddress = observation.captureAddress();
+
+		auto iter = CaptureAddressRangeIter(captureAddress);
+		if(iter != captureAddressRanges_.end()) {
+
+			const CaptureAddressRange captureAddressRange = iter->first;
+
+			const intptr_t offset = captureAddress.charPtr() -captureAddressRange.baseAddress().charPtr();
+			if(offset < captureAddressRange.length()) {
+				throw AddressAlreadyMappedException();
+			}
+		}
+
+		const size_t mappingLength = observation.resourceGenerator()->length();
+
+		char* replayAllocation = new char [mappingLength];
+		const ReplayAddress replayAddress(replayAllocation);
+
+		observation.resourceGenerator()->generate(replayAddress);
+
+		const CaptureAddressRange captureAddressRange = CaptureAddressRange(captureAddress, mappingLength);
+		const ReplayAddressRange replayAddressRange = ReplayAddressRange(replayAddress, mappingLength);
+
+		const auto newMapping = std::make_pair(captureAddressRange, replayAddressRange);
+		captureAddressRanges_.insert(newMapping);
+
+		return replayAddress;
+	}
+
+	void MemoryRemapper::RemoveMapping(const CaptureAddress& captureAddress)
+	{
+		auto iter = CaptureAddressRangeIter(captureAddress);
+		if(iter == captureAddressRanges_.end()) throw AddressNotMappedException();
+
+		const CaptureAddressRange captureAddressRange = iter->first;
+		if(captureAddressRange.baseAddress() != captureAddress) {
+
+			const intptr_t offset = captureAddress.charPtr() -captureAddressRange.baseAddress().charPtr();
+			if(offset >= captureAddressRange.length()) {
+				throw AddressNotMappedException();
+			}
+			else {
+				throw RemoveMappingOffsetAddressException();
+			}
+		}
+
+		const ReplayAddressRange replayAddressRange = iter->second;
+
+#ifndef NDEBUG
+		// In debug we'll splat all released memory with 0xDEAD before releasing it to help with debugging.
```

Comment:
Isn't this something most debugger does itself?

Also, I think we can separate this to a function + macro so it doesn't break the reading flow

Reply:
MSVC has this thing that whenever a memory gets allocated in a debug build, you would have it has `0xCCCCCCCC` or `0xCDCDCDCD`(stack or heap) and when it's freed `0xDDDDDDDD`. I would expect something similar for the Clang and GCC as well. So we can just directly check for those values.

My one concern is that, we probably forget to do this everywhere, therefore it will create a sense of false security and we shouldn't repeat what debuggers already do.

If we really need to do this, maybe we can do it as a macro so whenever we need this functionality we can just call the macro in a line which would be either noop or this operation based on build.

Reply:
I've moved this to a function. It's nice to have even if there are tools that can be used to achieve the same thing